src/generic_sbn_instance.hpp

```diff
@@ -0,0 +1,323 @@
+                                       site_pattern);
+  }
+
+  // Return a raw pointer to the engine if it's available.
+  Engine *GetEngine() const {
+    if (engine_ != nullptr) {
+      return engine_.get();
+    }
+    // else
+    Failwith(
+        "Engine not available. Call PrepareForPhyloLikelihood to make an "
+        "engine for phylogenetic likelihood computation computation.");
+  }
+
+  // Sample an integer index in [range.first, range.second) according to
+  // sbn_parameters_.
+  size_t SampleIndex(Range range) const {
+    const auto &[start, end] = range;
+    Assert(start < end && end <= sbn_parameters_.size(),
+           "SampleIndex given an invalid range.");
+    // We do not want to overwrite sbn_parameters so we make a copy.
+    EigenVectorXd sbn_parameters_subrange = sbn_parameters_.segment(start, end - start);
+    NumericalUtils::ProbabilityNormalizeInLog(sbn_parameters_subrange);
+    NumericalUtils::Exponentiate(sbn_parameters_subrange);
+    std::discrete_distribution<> distribution(sbn_parameters_subrange.begin(),
+                                              sbn_parameters_subrange.end());
+    // We have to add on range.first because we have taken a slice of the full
+    // array, and the sampler treats the beginning of this slice as zero.
+    auto result =
+        start + static_cast<size_t>(distribution(mersenne_twister_.GetGenerator()));
+    Assert(result < end, "SampleIndex sampled a value out of range.");
+    return result;
+  }
+
+  Node::NodePtr SampleTopology(bool rooted) const {
+    // Start by sampling a rootsplit.
+    size_t rootsplit_index =
+        SampleIndex(std::pair<size_t, size_t>(0, sbn_support_.RootsplitCount()));
+    const Bitset &rootsplit = sbn_support_.RootsplitsAt(rootsplit_index);
+    // The addition below turns the rootsplit into a subsplit.
+    auto topology = rooted ? SampleTopology(rootsplit + ~rootsplit)
+                           : SampleTopology(rootsplit + ~rootsplit)->Deroot();
+    topology->Polish();
+    return topology;
+  }
+
+  // The input to this function is a parent subsplit (of length 2n).
+  Node::NodePtr SampleTopology(const Bitset &parent_subsplit) const {
+    auto process_subsplit = [this](const Bitset &parent) {
+      auto singleton_option = parent.SplitChunk(1).SingletonOption();
+      if (singleton_option) {
+        return Node::Leaf(*singleton_option);
+      }  // else
+      auto child_index = SampleIndex(sbn_support_.ParentToRangeAt(parent));
+      return SampleTopology(sbn_support_.IndexToChildAt(child_index));
+    };
+    return Node::Join(process_subsplit(parent_subsplit),
+                      process_subsplit(parent_subsplit.RotateSubsplit()));
+  }
+
+  // Clear all of the state that depends on the current tree collection.
+  void ClearTreeCollectionAssociatedState() {
+    sbn_parameters_.resize(0);
+    topology_counter_.clear();
+    sbn_support_ = TSBNSupport();
```

Comment:
@ognian- style comments here? This is wiping out the previous `sbn_support_` by instead assigning a default-constructed one. This should activate the destructor, right? (I note that the "destructor" is a rule-of-zero one.)

Thank you for answering my newbie questions :)

Reply:
@matsen The destructor won't be called here, this calls the assignment operator. The destructor will be called only on the temporary object

Reply:
Oh!

Here is a simplified version to be more precise.

Let's say `Blah` is a class that takes a single `size_t` and makes a vector of floats out of it:

```cpp

struct Blah {
    Blah(size_t size) : my_vector_(std::vector(size)) {};
    std::vector<float> my_vector_
}

x = Blah(1)
x = Blah(2)
```
what happens to the vector from `Blah(1)` when we reassign?

Reply:
In this example vector's assignment operator is called, this happens for every member. No destructors are called until "x" goes out of scope.